components/zcash_address/src/encoding.rs

```diff
@@ -46,11 +46,19 @@ impl FromStr for ZcashAddress {
         let s = s.trim();
 
         // Try decoding as a unified address
-        if let Ok((net, data)) = unified::Address::decode(s) {
-            return Ok(ZcashAddress {
-                net,
-                kind: AddressKind::Unified(data),
-            });
+        match unified::Address::decode(s) {
+            Ok((net, data)) => {
+                return Ok(ZcashAddress {
+                    net,
+                    kind: AddressKind::Unified(data),
+                });
+            }
+            Err(unified::ParseError::NotUnified) => {
+                // allow decoding to fall through to Sapling/Transparent
+            }
+            Err(e) => {
+                return Err(ParseError::Unified(e));
```

Comment:
At this layer, we should be returning `ParseError::NotZcash` if the HRP is unknown. Replace this with:
```suggestion
                return Err(e.into());
```
and modify the `impl From<unified::ParseError> for ParseError` to map `UnknownPrefix` to `ParseError::NotZcash`.